### PR TITLE
[Events] Add subscribers automatically

### PR DESCRIPTION
**Pull request by spaghetticode**
**Description**

_(This PR is based on some work done by @elia and @AlessioRocco for [Meundies](https://github.com/meundies))_

Until now, adding subscribers to `Spree::Event.subscribers` was developers' responsibility.

This PR aims at automating the task by making Rails load subscriber files located in standard paths both in Solidus core and in the store app during the Rails initialization processs. When each file is required, it's now also added to the `Spree::Event.subscribers` list.

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [x] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)
- [ ] I have attached screenshots to this PR for visual changes (if needed)

### CODE REVIEW

File: core/lib/spree/event.rb [hunk 1]
```diff
@@ -26,6 +26,19 @@ def fire(event_name, opts = {})
...
     end
 
+    # Loads all Solidus' core and application's event subscribers files.
+    #
+    # Files must be placed under the directory `app/subscribers` and their name
+    # must end with `_subscriber.rb`.
+    def require_subscriber_files
+      pattern = "app/subscribers/**/*_subscriber.rb"
+      # Load Solidus subscribers
+      solidus_core_dir = Gem::Specification.find_by_name('solidus_core').gem_dir
```
[the remark this answers is not in the document]
Comment (spaghetticode) [reply core/lib/spree/event.rb]: See https://github.com/rubocop-hq/rubocop-rails/issues/51

File: guides/source/developers/events/overview.html.md [hunk 2]
```diff
@@ -94,10 +94,15 @@ method allows to map a method of the subscriber module to an event that
...
-you need to manually subscribe to them.
+These subscriber modules are loaded with the rest of your application, you
+don't need to manually subscribe them when:
 
-For example, you could subscribe to them programmatically with something like:
+* you put them in the directory (or any subdirectory of) `app/subscribers`
+* their filename ends with `_subscriber.rb`
+
+On the other hand, if you need to resort to manual subscription because you did
+not follow the naming convetion explained above, you can use an initializer
```
Comment (AlessioRocco) [inline guides/source/developers/events/overview.html.md:104]: Typo  `convetion` -> `convention`

**Comment by kennyadsl** [comment]
A couple of questions:

1. What about extensions? Do they need to register things manually?
2. What if someone already registered an event manually in their store? Does this end up with two subscriptions?

**Comment by spaghetticode** [comment]
@kennyadsl here you are my answers to your questions:

1) at this time extensions still need to register their events manually. I suppose we can achieve that by modifying `solidus_support`

2) event subscribers are subscribed only once, so doing it manually as well does create any problem. This is how `Spree::Event::Subscriber` already works, see [this spec](https://github.com/solidusio/solidus/blob/master/core/spec/lib/spree/event/subscriber_spec.rb#L41)

**Comment by aldesantis** [comment]
I'm not 100% sure about this to be honest, I think it adds a layer of magic that we may not need. It could be particularly confusing considering that core subscribers are loaded automatically, while all other codebases (extensions, the app etc.) will have to register their subscribers manually. At the end of the day, is it really worth it? 🤔

**Comment by spaghetticode** [comment]
@jarednorman I just checked, code reload still works fine in development (kudos go to @elia for the [implementation](https://github.com/solidusio/solidus/pull/3232)).

**Comment by aldesantis** [comment]
We discussed this during a core team meeting and, since the plan is to add the same capability to extensions, I'm happy to move forward with this! 🚀 Thanks @spaghetticode and @elia.

**Comment by jarednorman** [comment]
Cool. I'll hold off giving this an approval since it's still listed as "Needs Work" but I'm favour of this change overall.

**Comment by spaghetticode** [comment]
This functionality is made available on extensions by [this PR](https://github.com/solidusio/solidus_support/pull/45)  from `solidus_support`. 

Also, if anybody is interested in giving this feature a try I created a [demo app](https://github.com/spaghetticode/solidus_event_subscribers_autoload)  that quickly showcases event subscribers autoload from in Solidus core, Solidus extensions and the rails app itself.

**Comment by aldesantis** [comment]
@spaghetticode looks like https://github.com/solidusio/solidus/pull/3571/commits/f8f392a3fa739d41d4bef6216fa1c6f9fe61b9ff somehow ended up in this branch 😄 Can you remove it?

**Comment by spaghetticode** [comment]
@aldesantis 👍

**Comment by aldesantis** [comment]
@spaghetticode GTM! :shipit:

**Pull request merged**